xds/internal/balancer/cdsbalancer/cluster_handler.go

```diff
@@ -0,0 +1,185 @@
+	// A way to cancel the watch for the cluster.
+	cancelFunc func()
+
+	// A list of children, as the Node can be an aggregate Cluster.
+	children []*clusterNode
+
+	// A ClusterUpdate in order to build a list of cluster updates for CDS to send down to child
+	// XdsClusterResolverLoadBalancingPolicy.
+	clusterUpdate xdsclient.ClusterUpdate
+
+	// This boolean determines whether this Node has received the first update or not. This isn't the best practice,
+	// but this will protect a list of Cluster Updates from being constructed if a cluster in the tree has not received
+	// an update yet.
+	receivedFirstUpdate bool
+
+	clusterHandler *clusterHandler
+
+}
+
+// CreateClusterNode creates a cluster node from a given clusterName. This will also start the watch for that cluster.
+func createClusterNode(clusterName string, xdsClient xdsClientInterface, topLevelHandler *clusterHandler) *clusterNode {
+	c := &clusterNode{
+		clusterHandler: topLevelHandler,
+	}
+	// Communicate with the xds client here.
+	c.cancelFunc = xdsClient.WatchCluster(clusterName, c.handleResp)
+	return c
+}
+
+// This function serves as a wrapper around recursive logic in order to grab the lock.
+func (c *clusterNode) delete() {
+	c.clusterHandler.clusterMutex.Lock()
+	c.deleteRecursive()
+	c.clusterHandler.clusterMutex.Unlock()
+}
+
+func (c *clusterNode) deleteRecursive() {
+	c.cancelFunc()
+	for _, child := range c.children {
+		child.delete()
+	}
+}
+
+// Construct cluster update (potentially a list of ClusterUpdates) for a node.
+func (c *clusterNode) constructClusterUpdate() ([]xdsclient.ClusterUpdate, error) {
+	// If the cluster has not yet received an update, the cluster update is not yet ready.
+	if !c.receivedFirstUpdate {
+		return nil, errors.New("Tried to construct a cluster update on a cluster that has not received an update.")
+	}
+
+	// Base case - LogicalDNS or EDS. Both of these cluster types will be tied to a single ClusterUpdate.
+	if c.clusterUpdate.ClusterType != xdsclient.ClusterTypeAggregate {
+		return []xdsclient.ClusterUpdate{c.clusterUpdate}, nil
+	}
+
+	// If an aggregate construct a list by recursively calling down to all of it's children.
+	var childrenUpdates []xdsclient.ClusterUpdate
+	for _, child := range c.children {
+		childUpdateList, err := child.constructClusterUpdate()
+		if err != nil {
+			return nil, err
+		}
+		for _, clusterUpdate := range childUpdateList {
+			childrenUpdates = append(childrenUpdates, clusterUpdate)
+		}
+	}
+
+	return childrenUpdates, nil
+}
+
+// handleResp handles a xds response for a particular cluster. This function also
+// handles any logic with regards to any child state that may have changed.
+func (c *clusterNode) handleResp(clusterUpdate xdsclient.ClusterUpdate, err error) {
+	c.clusterHandler.clusterMutex.Lock()
+	if err != nil { // Write this error for run() to pick up in CDS LB policy.
+		c.clusterHandler.updateChannel.Put(&clusterHandlerUpdate{chu: nil, err: err})
+		return
```

Comment:
This return will leave `clusterMutex` locked, then any future lock will block.

An easy fix is to always defer unlock (though there are situations where you cannot defer, but those are rare)

```
c.clusterHandler.clusterMutex.Lock()
defer c.clusterHandler.clusterMutex.Unlock()
```

Reply:
Changed to defer.